parser/src/grammar/Production.java

```diff
@@ -1,75 +1,89 @@
 package grammar;
 
 import exceptions.GrammarException;
-
 import java.util.ArrayList;
 import java.util.List;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 public class Production {
 
-    public SymbolSequence sources;
-    public List<SymbolSequence> targets;
-
-    public Production(String productionString) {
-        List<String> productionTokens = ensureValidity(splitByArrow(productionString));
-
-        this.sources = extractSource(productionTokens.get(0));
-        this.targets = extractTargets(productionTokens.get(1));
-    }
-
-    public Production(String source, SymbolSequence target) {
-        this.sources = new SymbolSequence(source);
-        this.targets = new ArrayList<>();
-        this.targets.add(target);
+  public SymbolSequence sources;
+  public List<SymbolSequence> targets;
+  public Integer index;
+
+  public Production(String productionString) {
+    List<String> productionTokens = ensureValidity(splitByArrow(productionString));
+
+    this.sources = extractSource(productionTokens.get(0));
+    this.targets = extractTargets(productionTokens.get(1));
+  }
+
+  public Production(String source, SymbolSequence target) {
+    this.sources = new SymbolSequence(source);
+    this.targets = new ArrayList<>();
+    this.targets.add(target);
+    this.index = -1;
+  }
+
+  public Production(String source, SymbolSequence target, Integer index) {
+    this.sources = new SymbolSequence(source);
+    this.targets = new ArrayList<>();
+    this.targets.add(target);
+    this.index = index;
+  }
+
+  private List<String> splitByArrow(String productionString) {
```

Comment:
Would recommend generifying this method to split by any string given as a parameter